File: android/src/main/java/expo/module/signature/SignatureError.kt
package expo.module.signature

import java.security.InvalidAlgorithmParameterException
import java.security.InvalidKeyException
import java.security.KeyStoreException
import java.security.NoSuchAlgorithmException
import java.security.SignatureException
import java.security.UnrecoverableKeyException
import java.security.spec.InvalidParameterSpecException

enum class SignatureError(val code: String) {
    INVALID_PARAMETERS("INVALID_PARAMETERS"),
    INVALID_KEY("INVALID_KEY"),
    KEY_STORE_ERROR("KEY_STORE_ERROR"),
    NO_SUCH_ALGORITHM("NO_SUCH_ALGORITHM"),
    SIGNATURE_ERROR("SIGNATURE_ERROR"),
    KEY_EXPORT_ERROR("KEY_EXPORT_ERROR"),
    GENERAL_ERROR("GENERAL_ERROR");

    companion object {
        fun fromException(e: Exception): SignatureError {
            return when (e) {
                is InvalidAlgorithmParameterException, is InvalidParameterSpecException -> INVALID_PARAMETERS
                is InvalidKeyException -> INVALID_KEY
                is KeyStoreException, is UnrecoverableKeyException -> KEY_STORE_ERROR
                is NoSuchAlgorithmException -> NO_SUCH_ALGORITHM
                is SignatureException -> SIGNATURE_ERROR
                else -> GENERAL_ERROR
            }
        }
    }
}

File: ios/SignatureError.swift
//
//  SignatureError.swift
//  SignatureModule
//
//  Created by Riccardo Pizzoni on 11/07/24.
//

import ExpoModulesCore

enum SignatureError: String {
    case invalidParameters = "INVALID_PARAMETERS"
    case invalidKey = "INVALID_KEY"
    case keyStoreError = "KEY_STORE_ERROR"
    case noSuchAlgorithm = "NO_SUCH_ALGORITHM"
    case signatureError = "SIGNATURE_ERROR"
    case keyExportError = "KEY_EXPORT_ERROR"
    case generalError = "GENERAL_ERROR"
}

extension Exception {
    static func from(code: SignatureError, error: Error) -> Exception {
        return Exception(name: code.rawValue, description: error.localizedDescription).causedBy(error)
    }
    
    static func from(code: SignatureError, description: String) -> Exception {
        return Exception(name: code.rawValue, description: description)
    }
}
